apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,382 @@
+	//   - primary: Ascending order based on the value of the label key, interpreted as integers if present.
+	//   - secondary: Ascending order based on the name of the cluster if the label key is absent.
+	// +optional
+	SortingLabelKey *string `json:"sortingLabelKey,omitempty"`
+
+	// The collection of tasks that each stage needs to complete successfully before moving to the next stage.
+	// Each task is executed in parallel.
+	// +optional
+	PostStageTasks []PostStageTask `json:"postRolloutTasks,omitempty"`
+}
+
+// PostStageTask is the collection of post stage tasks that ALL need to be completed before we can move to the next stage.
+type PostStageTask struct {
+	// The type of the post stage task.
+	// +kubebuilder:validation:Enum=TimedWait;ManualApproval
+	// +required
+	Type PostStageTaskType `json:"type"`
+
+	// The name of the StagedUpdateApprovalRequest we need to create to get the approval for the next stage.
+	// We will generate a name with a prefix of the StagedUpdateRun name if it's empty.
+	// +optional
+	ApprovalRequestName string `json:"approvalTask,omitempty"`
+
+	// The time to wait after all the clusters in the current stage complete the update before we move to the next stage.
+	// +kubebuilder:default:="1h"
+	// +kubebuilder:validation:Pattern="^0|([0-9]+(\\.[0-9]+)?(s|m|h))+$"
+	// +kubebuilder:validation:Type:=string
+	// +optional
+	WaitTime metav1.Duration `json:"waitTime,omitempty"`
+}
+
+// StagedUpdateRunStatus defines the observed state of the StagedUpdateRun.
+type StagedUpdateRunStatus struct {
+	// ComputedStages list the name of the clusters in each stage when the updateRun started.
+	// The clusters in each stage are ordered following its order to be rolled out.
+	// The updateRun will stop if the cluster stages change during the update.
+	// +required
+	ComputedStages [][]string `json:"currentStages"`
+
+	// CurrentStageIndex is the index of the stage that is in the middle of the update.
+	// The index starts from 0. The index is -1 if the update is not started.
+	// The index is the size of the Stages if the update is completed.
+	// +required
+	CurrentStageIndex int `json:"currentStageIndex"`
+
+	// The applied strategy of the CRP that we are currently using.
+	// It is the same as the strategy in the CRP when we first start the update.
+	// We will NOT update the strategy in the update even if the appliedStrategy changes in the CRP.
+	// +required
+	CurrentAppliedStrategy string `json:"currentAppliedStrategy"`
+
+	// The collection of clusters on which we are going to remove the application from.
+	// We restart the update if CRP changes the selected clusters during our update.
+	// Therefore, we need to keep track of the clusters that we have applied the resources to but no long is selected.
+	// We remove the application from these clusters after all the stages are completed and remove them from this list.
+	// +required
+	ToBeRevertedClusters []string `json:"toBeRevertedClusters"`
```

Comment:
DeletionStageStatus StageStatus
contains a list of clusters whose work will be deleted ordered by the given sort label key.